theano/gpuarray/opt.py

```diff
@@ -376,9 +376,20 @@ def apply(self, fgraph):
             if new_o.type != o.type:
                 assert isinstance(o.type, tensor.TensorType)
                 assert isinstance(new_o.type, GpuArrayType)
-                new_o = host_from_gpu(new_o)
+                assert new_o.owner
+                # This condition is needed in the case one input is an
+                # output of the graph. Without this, it would
+                # introduce cycle as we don't replace correctly that
+                # case. It would also add extra transfer to/from the
+                # gpu.
+                if (isinstance(new_o.owner.op, GpuFromHost) and
```

Comment:
I remember we talking about this fix one day earlier. I was trying to see where the cycle is being formed. IIRC, there is one more place also. Are you able to remember ?